storage: TestReplicateRemovedNodeDisruptiveElection failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/6a236533824c797948eaddbd6bedf5e12eee1a13

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestReplicateRemovedNodeDisruptiveElection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=943249&tab=buildLog

```
I181004 07:42:13.011415 70339 storage/replica_command.go:812  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:/M{in-ax} [(n1,s1):1, next=2, gen=0]
I181004 07:42:13.021972 70339 storage/replica.go:3822  [s1,r1/1:/M{in-ax},txn=cff2c9fb] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181004 07:42:13.031716 70339 rpc/nodedialer/nodedialer.go:92  [s1,r1/1:/M{in-ax}] connection to n3 established
I181004 07:42:13.035131 70339 storage/store_snapshot.go:615  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 104e401b at applied index 19
I181004 07:42:13.036858 70339 storage/store_snapshot.go:657  [s1,r1/1:/M{in-ax}] streamed snapshot to (n3,s3):?: kv pairs: 53, log entries: 9, rate-limit: 2.0 MiB/sec, 4ms
I181004 07:42:13.039517 70329 storage/replica_raftstorage.go:803  [s3,r1/?:{-}] applying preemptive snapshot at index 19 (id=104e401b, encoded size=9460, 1 rocksdb batches, 9 log entries)
I181004 07:42:13.043477 70329 storage/replica_raftstorage.go:809  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 4ms [clear=0ms batch=0ms entries=3ms commit=0ms]
I181004 07:42:13.046463 70339 storage/replica_command.go:812  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I181004 07:42:13.055473 70431 rpc/nodedialer/nodedialer.go:92  connection to n1 established
I181004 07:42:13.067145 70339 storage/replica.go:3822  [s1,r1/1:/M{in-ax},txn=5bc1d6f9] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I181004 07:42:13.079657 70339 rpc/nodedialer/nodedialer.go:92  [s1,r1/1:/M{in-ax}] connection to n4 established
I181004 07:42:13.082389 70339 storage/store_snapshot.go:615  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 01276553 at applied index 21
I181004 07:42:13.083936 70339 storage/store_snapshot.go:657  [s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 56, log entries: 11, rate-limit: 2.0 MiB/sec, 4ms
I181004 07:42:13.086172 70671 storage/replica_raftstorage.go:803  [s4,r1/?:{-}] applying preemptive snapshot at index 21 (id=01276553, encoded size=10467, 1 rocksdb batches, 11 log entries)
I181004 07:42:13.090023 70671 storage/replica_raftstorage.go:809  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 4ms [clear=0ms batch=0ms entries=2ms commit=0ms]
I181004 07:42:13.092952 70339 storage/replica_command.go:812  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0]
I181004 07:42:13.104366 70339 storage/replica.go:3822  [s1,r1/1:/M{in-ax},txn=dfd42df7] proposing ADD_REPLICA((n4,s4):4): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3 (n4,s4):4] next=5
I181004 07:42:13.443534 70339 storage/replica_command.go:812  [s2,r1/2:/M{in-ax}] change replicas (REMOVE_REPLICA (n1,s1):1): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=0]
I181004 07:42:13.480275 70339 storage/replica.go:3822  [s2,r1/2:/M{in-ax},txn=b65cc2bc] proposing REMOVE_REPLICA((n1,s1):1): updated=[(n4,s4):4 (n2,s2):2 (n3,s3):3] next=5
I181004 07:42:13.494399 70331 storage/store.go:3800  [s1,r1/1:/M{in-ax}] added to replica GC queue (peer suggestion)
I181004 07:42:13.497515 70796 storage/store.go:2741  [replicaGC,s1,r1/1:/M{in-ax}] removing replica r1/1
I181004 07:42:13.499247 70796 storage/replica.go:872  [replicaGC,s1,r1/1:/M{in-ax}] removed 49 (44+5) keys in 1ms [clear=1ms commit=0ms]
I181004 07:42:13.508448 70339 rpc/nodedialer/nodedialer.go:92  connection to n2 established
I181004 07:42:14.840556 70621 gossip/gossip.go:1510  [n3] node has connected to cluster via gossip
I181004 07:42:14.897699 70780 gossip/gossip.go:1510  [n4] node has connected to cluster via gossip
I181004 07:42:15.089800 70532 gossip/gossip.go:1510  [n2] node has connected to cluster via gossip
I181004 07:42:28.501025 70420 storage/compactor/compactor.go:330  [s1,compactor] purging suggested compaction for range /Min - /Max that contains live data
--- FAIL: TestReplicateRemovedNodeDisruptiveElection (45.88s)
	client_raft_test.go:3212: did not get expected error

```

SHA: https://github.com/cockroachdb/cockroach/commits/32346197937265bbc71dd3629b1137df9bf16fb4

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestReplicateRemovedNodeDisruptiveElection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=948435&tab=buildLog

```
I181007 05:30:07.699425 71338 storage/replica_command.go:812  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:/M{in-ax} [(n1,s1):1, next=2, gen=0]
I181007 05:30:07.701059 71338 storage/replica.go:3825  [s1,r1/1:/M{in-ax},txn=bb83a35f] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181007 05:30:07.702987 71338 storage/store_snapshot.go:615  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 375d6b94 at applied index 19
I181007 05:30:07.703440 71338 storage/store_snapshot.go:657  [s1,r1/1:/M{in-ax}] streamed snapshot to (n3,s3):?: kv pairs: 53, log entries: 9, rate-limit: 2.0 MiB/sec, 1ms
I181007 05:30:07.703896 71885 storage/replica_raftstorage.go:803  [s3,r1/?:{-}] applying preemptive snapshot at index 19 (id=375d6b94, encoded size=9460, 1 rocksdb batches, 9 log entries)
I181007 05:30:07.710370 71885 storage/replica_raftstorage.go:809  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 5ms [clear=0ms batch=0ms entries=5ms commit=0ms]
I181007 05:30:07.711837 71338 storage/replica_command.go:812  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I181007 05:30:07.714410 71338 storage/replica.go:3825  [s1,r1/1:/M{in-ax},txn=b8e006b5] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I181007 05:30:07.716337 71338 storage/store_snapshot.go:615  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 939128bd at applied index 21
I181007 05:30:07.716612 71338 storage/store_snapshot.go:657  [s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 56, log entries: 11, rate-limit: 2.0 MiB/sec, 1ms
I181007 05:30:07.716884 71933 storage/replica_raftstorage.go:803  [s4,r1/?:{-}] applying preemptive snapshot at index 21 (id=939128bd, encoded size=10467, 1 rocksdb batches, 11 log entries)
I181007 05:30:07.717769 71933 storage/replica_raftstorage.go:809  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=1ms commit=0ms]
I181007 05:30:07.718226 71338 storage/replica_command.go:812  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0]
I181007 05:30:07.720218 71338 storage/replica.go:3825  [s1,r1/1:/M{in-ax},txn=b2f6f320] proposing ADD_REPLICA((n4,s4):4): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3 (n4,s4):4] next=5
I181007 05:30:07.868265 71338 storage/replica_command.go:812  [s2,r1/2:/M{in-ax}] change replicas (REMOVE_REPLICA (n1,s1):1): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=0]
I181007 05:30:07.870916 71338 storage/replica.go:3825  [s2,r1/2:/M{in-ax},txn=9dadd3dc] proposing REMOVE_REPLICA((n1,s1):1): updated=[(n4,s4):4 (n2,s2):2 (n3,s3):3] next=5
I181007 05:30:07.873779 71940 storage/store.go:2741  [replicaGC,s1,r1/1:/M{in-ax}] removing replica r1/1
I181007 05:30:07.874177 71940 storage/replica.go:875  [replicaGC,s1,r1/1:/M{in-ax}] removed 49 (44+5) keys in 0ms [clear=0ms commit=0ms]
I181007 05:30:09.370708 71757 gossip/gossip.go:1510  [n3] node has connected to cluster via gossip
I181007 05:30:09.856805 71651 gossip/gossip.go:1510  [n2] node has connected to cluster via gossip
I181007 05:30:10.023693 71323 gossip/gossip.go:1510  [n4] node has connected to cluster via gossip
I181007 05:30:22.874460 71529 storage/compactor/compactor.go:330  [s1,compactor] purging suggested compaction for range /Min - /Max that contains live data
W181007 05:30:52.875232 71888 storage/raft_transport.go:583  while processing outgoing Raft queue to node 2: EOF:
W181007 05:30:52.875240 71934 storage/raft_transport.go:583  while processing outgoing Raft queue to node 3: EOF:
W181007 05:30:52.875233 71935 storage/raft_transport.go:583  while processing outgoing Raft queue to node 4: EOF:
W181007 05:30:52.875254 71866 storage/raft_transport.go:583  while processing outgoing Raft queue to node 2: EOF:
--- FAIL: TestReplicateRemovedNodeDisruptiveElection (45.21s)
	client_raft_test.go:3217: did not get expected ReplicaTooOldError error

```

SHA: https://github.com/cockroachdb/cockroach/commits/7aaff768f983cf142001e5ee6eaf7262913b8985

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestReplicateRemovedNodeDisruptiveElection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1006281&tab=buildLog

```
I181110 08:49:51.775052 72677 storage/replica_raftstorage.go:803  [s2,r1/?:{-}] applying preemptive snapshot at index 17 (id=044db7f6, encoded size=8476, 1 rocksdb batches, 7 log entries)
I181110 08:49:51.779450 72677 storage/replica_raftstorage.go:809  [s2,r1/?:/M{in-ax}] applied preemptive snapshot in 4ms [clear=0ms batch=0ms entries=3ms commit=0ms]
I181110 08:49:51.783569 72693 storage/replica_command.go:814  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:/M{in-ax} [(n1,s1):1, next=2, gen=0]
I181110 08:49:51.795244 72693 storage/replica.go:3838  [s1,r1/1:/M{in-ax},txn=dae7b174] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181110 08:49:51.814099 72693 storage/store_snapshot.go:636  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 9063ae19 at applied index 19
I181110 08:49:51.815958 72693 storage/store_snapshot.go:679  [s1,r1/1:/M{in-ax}] streamed snapshot to (n3,s3):?: kv pairs: 53, log entries: 9, rate-limit: 2.0 MiB/sec, 5ms
I181110 08:49:51.817991 72810 storage/replica_raftstorage.go:803  [s3,r1/?:{-}] applying preemptive snapshot at index 19 (id=9063ae19, encoded size=9418, 1 rocksdb batches, 9 log entries)
I181110 08:49:51.826202 72810 storage/replica_raftstorage.go:809  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 8ms [clear=0ms batch=0ms entries=7ms commit=0ms]
I181110 08:49:51.831993 72693 storage/replica_command.go:814  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I181110 08:49:51.860308 72693 storage/replica.go:3838  [s1,r1/1:/M{in-ax},txn=4e75b5d1] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I181110 08:49:51.881981 72693 storage/store_snapshot.go:636  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 75849f95 at applied index 21
I181110 08:49:51.886601 73098 storage/replica_raftstorage.go:803  [s4,r1/?:{-}] applying preemptive snapshot at index 21 (id=75849f95, encoded size=10425, 1 rocksdb batches, 11 log entries)
I181110 08:49:51.897495 73098 storage/replica_raftstorage.go:809  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 11ms [clear=0ms batch=0ms entries=9ms commit=0ms]
I181110 08:49:51.898926 72693 storage/store_snapshot.go:679  [s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 56, log entries: 11, rate-limit: 2.0 MiB/sec, 19ms
I181110 08:49:51.903876 72693 storage/replica_command.go:814  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0]
I181110 08:49:51.946759 72693 storage/replica.go:3838  [s1,r1/1:/M{in-ax},txn=7e8e79ef] proposing ADD_REPLICA((n4,s4):4): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3 (n4,s4):4] next=5
I181110 08:49:52.065928 72693 storage/replica_command.go:814  [s2,r1/2:/M{in-ax}] change replicas (REMOVE_REPLICA (n1,s1):1): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=0]
I181110 08:49:52.119028 72693 storage/replica.go:3838  [s2,r1/2:/M{in-ax},txn=a32b557a] proposing REMOVE_REPLICA((n1,s1):1): updated=[(n4,s4):4 (n2,s2):2 (n3,s3):3] next=5
I181110 08:49:52.145641 73157 storage/store.go:3785  [s1,r1/1:/M{in-ax}] added to replica GC queue (peer suggestion)
I181110 08:49:52.159087 73104 storage/store.go:2725  [replicaGC,s1,r1/1:/M{in-ax}] removing replica r1/1
I181110 08:49:52.165788 73104 storage/replica.go:863  [replicaGC,s1,r1/1:/M{in-ax}] removed 49 (44+5) keys in 5ms [clear=4ms commit=0ms]
I181110 08:49:53.142105 73026 gossip/gossip.go:1513  [n3] node has connected to cluster via gossip
I181110 08:49:53.142854 72552 gossip/gossip.go:1513  [n2] node has connected to cluster via gossip
I181110 08:49:53.264474 73140 gossip/gossip.go:1513  [n4] node has connected to cluster via gossip
I181110 08:50:07.166595 72774 storage/compactor/compactor.go:330  [s1,compactor] purging suggested compaction for range /Min - /Max that contains live data
--- FAIL: TestReplicateRemovedNodeDisruptiveElection (45.91s)
	client_raft_test.go:3212: did not get expected error

```

SHA: https://github.com/cockroachdb/cockroach/commits/859214b81838a4ba33048b81497442ce5774baa7

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestReplicateRemovedNodeDisruptiveElection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1053459&tab=buildLog

```
I181213 06:41:56.307552 61870 storage/replica_command.go:899  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:/M{in-ax} [(n1,s1):1, next=2, gen=0]
I181213 06:41:56.309043 61870 storage/replica.go:3817  [s1,r1/1:/M{in-ax},txn=d6771bda] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181213 06:41:56.312158 61870 storage/store_snapshot.go:760  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 79e72fd8 at applied index 19
I181213 06:41:56.312575 61870 storage/store_snapshot.go:803  [s1,r1/1:/M{in-ax}] streamed snapshot to (n3,s3):?: kv pairs: 53, log entries: 9, rate-limit: 2.0 MiB/sec, 0.00s
I181213 06:41:56.312953 62209 storage/replica_raftstorage.go:809  [s3,r1/?:{-}] applying preemptive snapshot at index 19 (id=79e72fd8, encoded size=9462, 1 rocksdb batches, 9 log entries)
I181213 06:41:56.314132 62209 storage/replica_raftstorage.go:815  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=1ms commit=0ms]
I181213 06:41:56.314993 61870 storage/replica_command.go:899  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I181213 06:41:56.318164 61870 storage/replica.go:3817  [s1,r1/1:/M{in-ax},txn=70e8d4f2] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I181213 06:41:56.320026 61870 storage/store_snapshot.go:760  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 8e941002 at applied index 21
I181213 06:41:56.320334 61870 storage/store_snapshot.go:803  [s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 56, log entries: 11, rate-limit: 2.0 MiB/sec, 0.00s
I181213 06:41:56.320655 62500 storage/replica_raftstorage.go:809  [s4,r1/?:{-}] applying preemptive snapshot at index 21 (id=8e941002, encoded size=10469, 1 rocksdb batches, 11 log entries)
I181213 06:41:56.321533 62500 storage/replica_raftstorage.go:815  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=1ms commit=0ms]
I181213 06:41:56.321939 61870 storage/replica_command.go:899  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0]
I181213 06:41:56.323791 61870 storage/replica.go:3817  [s1,r1/1:/M{in-ax},txn=37f82996] proposing ADD_REPLICA((n4,s4):4): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3 (n4,s4):4] next=5
I181213 06:41:56.465053 61870 storage/replica_command.go:899  [s2,r1/2:/M{in-ax}] change replicas (REMOVE_REPLICA (n1,s1):1): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=0]
I181213 06:41:56.467297 61870 storage/replica.go:3817  [s2,r1/2:/M{in-ax},txn=4e8d55dd] proposing REMOVE_REPLICA((n1,s1):1): updated=[(n4,s4):4 (n2,s2):2 (n3,s3):3] next=5
I181213 06:41:56.470270 62476 storage/store.go:2561  [replicaGC,s1,r1/1:/M{in-ax}] removing replica r1/1
I181213 06:41:56.470672 62476 storage/replica.go:908  [replicaGC,s1,r1/1:/M{in-ax}] removed 49 (44+5) keys in 0ms [clear=0ms commit=0ms]
I181213 06:41:57.923787 61889 gossip/gossip.go:1513  [n2] node has connected to cluster via gossip
I181213 06:41:58.171314 62330 gossip/gossip.go:1513  [n3] node has connected to cluster via gossip
I181213 06:41:58.356414 62439 gossip/gossip.go:1513  [n4] node has connected to cluster via gossip
I181213 06:42:11.470998 62047 storage/compactor/compactor.go:329  [s1,compactor] purging suggested compaction for range /Min - /Max that contains live data
W181213 06:42:41.471509 62439 gossip/gossip.go:1499  [n4] no incoming or outgoing connections
W181213 06:42:41.471600 62330 gossip/gossip.go:1499  [n3] no incoming or outgoing connections
W181213 06:42:41.471656 62314 storage/raft_transport.go:583  while processing outgoing Raft queue to node 4: rpc error: code = Unavailable desc = transport is closing:
--- FAIL: TestReplicateRemovedNodeDisruptiveElection (45.30s)
	client_raft_test.go:3252: did not get expected ReplicaTooOldError error

```

SHA: https://github.com/cockroachdb/cockroach/commits/e7dc507fa0ecc7dc5ed597ca5c6cdeb48086428c

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestReplicateRemovedNodeDisruptiveElection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1057785&tab=buildLog

```
I181216 09:08:59.550862 61852 storage/replica_raftstorage.go:809  [s2,r1/?:{-}] applying preemptive snapshot at index 17 (id=d481098d, encoded size=8560, 1 rocksdb batches, 7 log entries)
I181216 09:08:59.554170 61852 storage/replica_raftstorage.go:815  [s2,r1/?:/M{in-ax}] applied preemptive snapshot in 3ms [clear=0ms batch=0ms entries=2ms commit=0ms]
I181216 09:08:59.559661 61397 storage/replica_command.go:899  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:/M{in-ax} [(n1,s1):1, next=2, gen=0]
I181216 09:08:59.567834 61397 storage/replica.go:3823  [s1,r1/1:/M{in-ax},txn=7c9dc262] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181216 09:08:59.579632 61397 storage/store_snapshot.go:762  [s1,r1/1:/M{in-ax}] sending preemptive snapshot e4005c05 at applied index 19
I181216 09:08:59.581101 61397 storage/store_snapshot.go:805  [s1,r1/1:/M{in-ax}] streamed snapshot to (n3,s3):?: kv pairs: 53, log entries: 9, rate-limit: 2.0 MiB/sec, 0.00s
I181216 09:08:59.583291 61864 storage/replica_raftstorage.go:809  [s3,r1/?:{-}] applying preemptive snapshot at index 19 (id=e4005c05, encoded size=9504, 1 rocksdb batches, 9 log entries)
I181216 09:08:59.587352 61864 storage/replica_raftstorage.go:815  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 4ms [clear=0ms batch=0ms entries=2ms commit=0ms]
I181216 09:08:59.590869 61397 storage/replica_command.go:899  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I181216 09:08:59.615220 61397 storage/replica.go:3823  [s1,r1/1:/M{in-ax},txn=a3b3a851] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I181216 09:08:59.636614 61397 storage/store_snapshot.go:762  [s1,r1/1:/M{in-ax}] sending preemptive snapshot bc304f52 at applied index 21
I181216 09:08:59.638592 61397 storage/store_snapshot.go:805  [s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 56, log entries: 11, rate-limit: 2.0 MiB/sec, 0.00s
I181216 09:08:59.640766 61878 storage/replica_raftstorage.go:809  [s4,r1/?:{-}] applying preemptive snapshot at index 21 (id=bc304f52, encoded size=10513, 1 rocksdb batches, 11 log entries)
I181216 09:08:59.645231 61878 storage/replica_raftstorage.go:815  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 4ms [clear=0ms batch=0ms entries=3ms commit=0ms]
I181216 09:08:59.650668 61397 storage/replica_command.go:899  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0]
I181216 09:08:59.664454 61397 storage/replica.go:3823  [s1,r1/1:/M{in-ax},txn=08768343] proposing ADD_REPLICA((n4,s4):4): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3 (n4,s4):4] next=5
I181216 09:09:00.134145 61397 storage/replica_command.go:899  [s2,r1/2:/M{in-ax}] change replicas (REMOVE_REPLICA (n1,s1):1): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=0]
I181216 09:09:00.149881 61397 storage/replica.go:3823  [s2,r1/2:/M{in-ax},txn=cc54ba6a] proposing REMOVE_REPLICA((n1,s1):1): updated=[(n4,s4):4 (n2,s2):2 (n3,s3):3] next=5
I181216 09:09:00.179481 61923 storage/store.go:2561  [replicaGC,s1,r1/1:/M{in-ax}] removing replica r1/1
I181216 09:09:00.181359 61923 storage/replica.go:914  [replicaGC,s1,r1/1:/M{in-ax}] removed 49 (44+5) keys in 1ms [clear=1ms commit=0ms]
I181216 09:09:01.152028 61506 gossip/gossip.go:1513  [n3] node has connected to cluster via gossip
I181216 09:09:01.275786 61846 gossip/gossip.go:1513  [n4] node has connected to cluster via gossip
I181216 09:09:01.496951 61501 gossip/gossip.go:1513  [n2] node has connected to cluster via gossip
I181216 09:09:15.183799 61488 storage/compactor/compactor.go:329  [s1,compactor] purging suggested compaction for range /Min - /Max that contains live data
W181216 09:09:45.257790 61506 gossip/gossip.go:1499  [n3] no incoming or outgoing connections
--- FAIL: TestReplicateRemovedNodeDisruptiveElection (45.96s)
	client_raft_test.go:3252: did not get expected ReplicaTooOldError error

```

SHA: https://github.com/cockroachdb/cockroach/commits/a119a3a158725c9e3f9b8084d9398601c0e67007

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestReplicateRemovedNodeDisruptiveElection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1170975&tab=buildLog

```
I190311 05:37:08.197215 646974 storage/replica_command.go:788  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I190311 05:37:08.200124 646974 storage/replica_raft.go:372  [s1,r1/1:/M{in-ax},txn=58adeadc] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I190311 05:37:08.202529 646974 storage/store_snapshot.go:775  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 94ec08bc at applied index 21
I190311 05:37:08.202777 646974 storage/store_snapshot.go:818  [s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 60, log entries: 11, rate-limit: 8.0 MiB/sec, 0.00s
I190311 05:37:08.203207 648323 storage/replica_raftstorage.go:814  [s4,r1/?:{-}] applying preemptive snapshot at index 21 (id=94ec08bc, encoded size=10699, 1 rocksdb batches, 11 log entries)
I190311 05:37:08.209705 648323 storage/replica_raftstorage.go:820  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 6ms [clear=0ms batch=0ms entries=1ms commit=5ms]
I190311 05:37:08.215804 646974 storage/replica_command.go:788  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0]
I190311 05:37:08.221409 646974 storage/replica_raft.go:372  [s1,r1/1:/M{in-ax},txn=c5b72b26] proposing ADD_REPLICA((n4,s4):4): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3 (n4,s4):4] next=5
I190311 05:37:08.536252 646974 storage/replica_command.go:788  [s2,r1/2:/M{in-ax}] change replicas (REMOVE_REPLICA (n1,s1):1): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=0]
I190311 05:37:08.539185 646974 storage/replica_raft.go:372  [s2,r1/2:/M{in-ax},txn=bf262446] proposing REMOVE_REPLICA((n1,s1):1): updated=[(n4,s4):4 (n2,s2):2 (n3,s3):3] next=5
I190311 05:37:08.553210 649404 storage/store.go:2453  [replicaGC,s1,r1/1:/M{in-ax}] removing replica r1/1
I190311 05:37:08.554311 649404 storage/replica_destroy.go:150  [replicaGC,s1,r1/1:/M{in-ax}] removed 53 (49+4) keys in 1ms [clear=1ms commit=0ms]
I190311 05:37:09.747656 647314 gossip/gossip.go:1510  [n2] node has connected to cluster via gossip
I190311 05:37:10.211537 647531 gossip/gossip.go:1510  [n3] node has connected to cluster via gossip
I190311 05:37:10.523802 647942 gossip/gossip.go:1510  [n4] node has connected to cluster via gossip
I190311 05:37:23.566904 647233 storage/compactor/compactor.go:329  [s1,compactor] purging suggested compaction for range /Min - /Max that contains live data
I190311 05:37:53.553128 843838 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190311 05:37:53.553163 843838 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190311 05:37:53.553185 843835 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190311 05:37:53.553202 843835 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190311 05:37:53.553227 843836 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190311 05:37:53.553244 843836 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190311 05:37:53.553271 843837 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190311 05:37:53.553288 843837 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
W190311 05:37:53.557762 648355 storage/raft_transport.go:583  while processing outgoing Raft queue to node 3: rpc error: code = Unavailable desc = transport is closing:
--- FAIL: TestReplicateRemovedNodeDisruptiveElection (45.62s)
    client_raft_test.go:3284: did not get expected ReplicaTooOldError error

```

SHA: https://github.com/cockroachdb/cockroach/commits/5b36cc6276340282cb333ff4a9cb4f1fbd6c3348

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stress TESTS=TestReplicateRemovedNodeDisruptiveElection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1190443&tab=buildLog

```
I190321 06:50:29.173020 73992 storage/replica_raftstorage.go:818  [s2,r1/?:{-}] applying preemptive snapshot at index 17 (id=ddcfe89b, encoded size=8478, 1 rocksdb batches, 7 log entries)
I190321 06:50:29.176181 73992 storage/replica_raftstorage.go:824  [s2,r1/?:/M{in-ax}] applied preemptive snapshot in 3ms [clear=0ms batch=0ms entries=2ms commit=0ms]
I190321 06:50:29.179612 73521 storage/replica_command.go:898  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:/M{in-ax} [(n1,s1):1, next=2, gen=0]
I190321 06:50:29.187916 73521 storage/replica.go:3883  [s1,r1/1:/M{in-ax},txn=12cf7ef1] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I190321 06:50:29.209207 73521 storage/store_snapshot.go:655  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 2fb06cce at applied index 19
I190321 06:50:29.210589 73521 storage/store_snapshot.go:698  [s1,r1/1:/M{in-ax}] streamed snapshot to (n3,s3):?: kv pairs: 53, log entries: 9, rate-limit: 2.0 MiB/sec, 0.00s
I190321 06:50:29.212134 74137 storage/replica_raftstorage.go:818  [s3,r1/?:{-}] applying preemptive snapshot at index 19 (id=2fb06cce, encoded size=9422, 1 rocksdb batches, 9 log entries)
I190321 06:50:29.215541 74137 storage/replica_raftstorage.go:824  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 3ms [clear=0ms batch=0ms entries=2ms commit=0ms]
I190321 06:50:29.218930 73521 storage/replica_command.go:898  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I190321 06:50:29.242052 73521 storage/replica.go:3883  [s1,r1/1:/M{in-ax},txn=1cd1644e] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I190321 06:50:29.258720 73521 storage/store_snapshot.go:655  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 7f105c64 at applied index 21
I190321 06:50:29.260312 73521 storage/store_snapshot.go:698  [s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 56, log entries: 11, rate-limit: 2.0 MiB/sec, 0.00s
I190321 06:50:29.261753 73650 storage/replica_raftstorage.go:818  [s4,r1/?:{-}] applying preemptive snapshot at index 21 (id=7f105c64, encoded size=10431, 1 rocksdb batches, 11 log entries)
I190321 06:50:29.265786 73650 storage/replica_raftstorage.go:824  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 4ms [clear=0ms batch=0ms entries=3ms commit=0ms]
I190321 06:50:29.271466 73521 storage/replica_command.go:898  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0]
I190321 06:50:29.285185 73521 storage/replica.go:3883  [s1,r1/1:/M{in-ax},txn=efd17b8f] proposing ADD_REPLICA((n4,s4):4): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3 (n4,s4):4] next=5
I190321 06:50:29.615671 73521 storage/replica_command.go:898  [s2,r1/2:/M{in-ax}] change replicas (REMOVE_REPLICA (n1,s1):1): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=0]
I190321 06:50:29.685235 73521 storage/replica.go:3883  [s2,r1/2:/M{in-ax},txn=2f296d91] proposing REMOVE_REPLICA((n1,s1):1): updated=[(n4,s4):4 (n2,s2):2 (n3,s3):3] next=5
I190321 06:50:29.696603 73997 storage/store.go:3806  [s1,r1/1:/M{in-ax}] added to replica GC queue (peer suggestion)
I190321 06:50:29.701760 74195 storage/store.go:2742  [replicaGC,s1,r1/1:/M{in-ax}] removing replica r1/1
I190321 06:50:29.703928 74195 storage/replica.go:941  [replicaGC,s1,r1/1:/M{in-ax}] removed 49 (44+5) keys in 1ms [clear=1ms commit=0ms]
I190321 06:50:30.654714 73875 gossip/gossip.go:1489  [n2] node has connected to cluster via gossip
I190321 06:50:30.809036 73759 gossip/gossip.go:1489  [n3] node has connected to cluster via gossip
I190321 06:50:30.876686 74122 gossip/gossip.go:1489  [n4] node has connected to cluster via gossip
I190321 06:50:44.706058 73730 storage/compactor/compactor.go:330  [s1,compactor] purging suggested compaction for range /Min - /Max that contains live data
--- FAIL: TestReplicateRemovedNodeDisruptiveElection (45.86s)
	client_raft_test.go:3229: did not get expected error

```

SHA: https://github.com/cockroachdb/cockroach/commits/a6b3c540b696002b2ed07036a657612995d6d1ab

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestReplicateRemovedNodeDisruptiveElection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1211542&tab=buildLog

```
I190331 06:05:50.283381 517253 storage/replica_raftstorage.go:820  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190331 06:05:50.283764 516120 storage/replica_command.go:796  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I190331 06:05:50.303473 516120 storage/replica_raft.go:394  [s1,r1/1:/M{in-ax},txn=fa828823] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I190331 06:05:50.335997 516120 storage/store_snapshot.go:775  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 2cfef4ea at applied index 21
I190331 06:05:50.336323 516120 storage/store_snapshot.go:818  [s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 60, log entries: 0, rate-limit: 8.0 MiB/sec, 0.03s
I190331 06:05:50.336697 517149 storage/replica_raftstorage.go:814  [s4,r1/?:{-}] applying preemptive snapshot at index 21 (id=2cfef4ea, encoded size=8515, 1 rocksdb batches, 0 log entries)
I190331 06:05:50.336952 517149 storage/replica_raftstorage.go:820  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190331 06:05:50.337376 516120 storage/replica_command.go:796  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0]
I190331 06:05:50.339159 516120 storage/replica_raft.go:394  [s1,r1/1:/M{in-ax},txn=c009bac2] proposing ADD_REPLICA((n4,s4):4): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3 (n4,s4):4] next=5
I190331 06:05:50.798367 516120 storage/replica_command.go:796  [s2,r1/2:/M{in-ax}] change replicas (REMOVE_REPLICA (n1,s1):1): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=0]
I190331 06:05:50.800294 516120 storage/replica_raft.go:394  [s2,r1/2:/M{in-ax},txn=fac549b8] proposing REMOVE_REPLICA((n1,s1):1): updated=[(n4,s4):4 (n2,s2):2 (n3,s3):3] next=5
I190331 06:05:50.820910 518835 storage/store.go:2479  [replicaGC,s1,r1/1:/M{in-ax}] removing replica r1/1
I190331 06:05:50.821443 518835 storage/replica_destroy.go:150  [replicaGC,s1,r1/1:/M{in-ax}] removed 53 (49+4) keys in 0ms [clear=0ms commit=0ms]
I190331 06:05:51.797156 516612 gossip/gossip.go:1510  [n3] node has connected to cluster via gossip
I190331 06:05:52.049554 516876 gossip/gossip.go:1510  [n4] node has connected to cluster via gossip
I190331 06:05:52.259839 516002 gossip/gossip.go:1510  [n2] node has connected to cluster via gossip
I190331 06:06:05.826014 516203 storage/compactor/compactor.go:329  [s1,compactor] purging suggested compaction for range /Min - /Max that contains live data
I190331 06:06:35.804761 738822 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190331 06:06:35.804806 738821 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190331 06:06:35.804819 738822 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190331 06:06:35.804847 738821 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190331 06:06:35.805094 738819 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190331 06:06:35.805131 738819 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190331 06:06:35.805269 738820 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190331 06:06:35.805305 738820 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_gc_batcher
--- FAIL: TestReplicateRemovedNodeDisruptiveElection (45.75s)
    client_raft_test.go:3285: did not get expected ReplicaTooOldError error

```

SHA: https://github.com/cockroachdb/cockroach/commits/a039a93a5cc6eb3f395ceb6f7dc8030985dccc29

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestReplicateRemovedNodeDisruptiveElection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1212538&tab=buildLog

```
I190401 05:32:23.024078 684290 storage/store_snapshot.go:818  [s1,r1/1:/M{in-ax}] streamed snapshot to (n3,s3):?: kv pairs: 57, log entries: 0, rate-limit: 8.0 MiB/sec, 0.00s
I190401 05:32:23.024510 685513 storage/replica_raftstorage.go:814  [s3,r1/?:{-}] applying preemptive snapshot at index 19 (id=36a04202, encoded size=8516, 1 rocksdb batches, 0 log entries)
I190401 05:32:23.024799 685513 storage/replica_raftstorage.go:820  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190401 05:32:23.025186 684290 storage/replica_command.go:796  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I190401 05:32:23.028538 684290 storage/replica_raft.go:394  [s1,r1/1:/M{in-ax},txn=baeb8a46] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I190401 05:32:23.031253 684290 storage/store_snapshot.go:775  [s1,r1/1:/M{in-ax}] sending preemptive snapshot b0d1d551 at applied index 21
I190401 05:32:23.031466 684290 storage/store_snapshot.go:818  [s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 60, log entries: 0, rate-limit: 8.0 MiB/sec, 0.00s
I190401 05:32:23.031835 685571 storage/replica_raftstorage.go:814  [s4,r1/?:{-}] applying preemptive snapshot at index 21 (id=b0d1d551, encoded size=8699, 1 rocksdb batches, 0 log entries)
I190401 05:32:23.032142 685571 storage/replica_raftstorage.go:820  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190401 05:32:23.032499 684290 storage/replica_command.go:796  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0]
I190401 05:32:23.035504 684290 storage/replica_raft.go:394  [s1,r1/1:/M{in-ax},txn=eb6788b7] proposing ADD_REPLICA((n4,s4):4): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3 (n4,s4):4] next=5
I190401 05:32:23.340352 684290 storage/replica_command.go:796  [s2,r1/2:/M{in-ax}] change replicas (REMOVE_REPLICA (n1,s1):1): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=0]
I190401 05:32:23.343654 684290 storage/replica_raft.go:394  [s2,r1/2:/M{in-ax},txn=cda499e0] proposing REMOVE_REPLICA((n1,s1):1): updated=[(n4,s4):4 (n2,s2):2 (n3,s3):3] next=5
I190401 05:32:23.346542 686400 storage/store.go:2471  [replicaGC,s1,r1/1:/M{in-ax}] removing replica r1/1
I190401 05:32:23.346980 686400 storage/replica_destroy.go:150  [replicaGC,s1,r1/1:/M{in-ax}] removed 53 (49+4) keys in 0ms [clear=0ms commit=0ms]
I190401 05:32:24.433867 684556 gossip/gossip.go:1510  [n2] node has connected to cluster via gossip
I190401 05:32:24.473489 685089 gossip/gossip.go:1510  [n4] node has connected to cluster via gossip
I190401 05:32:24.474762 684826 gossip/gossip.go:1510  [n3] node has connected to cluster via gossip
I190401 05:32:38.347295 684549 storage/compactor/compactor.go:329  [s1,compactor] purging suggested compaction for range /Min - /Max that contains live data
I190401 05:33:08.347383 919714 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190401 05:33:08.347383 919713 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190401 05:33:08.347425 919714 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_gc_batcher
I190401 05:33:08.347441 919711 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190401 05:33:08.347449 919712 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
W190401 05:33:08.348209 684556 gossip/gossip.go:1496  [n2] no incoming or outgoing connections
--- FAIL: TestReplicateRemovedNodeDisruptiveElection (45.58s)
    client_raft_test.go:3282: did not get expected ReplicaTooOldError error

```

SHA: https://github.com/cockroachdb/cockroach/commits/2851c7d56ee4966109691b5c48c73ec8d4cc9847

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestReplicateRemovedNodeDisruptiveElection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1215555&tab=buildLog

```
I190402 05:44:02.912120 401623 storage/replica_raftstorage.go:820  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190402 05:44:02.912546 400031 storage/replica_command.go:796  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I190402 05:44:02.915935 400031 storage/replica_raft.go:394  [s1,r1/1:/M{in-ax},txn=091b4519] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I190402 05:44:02.925521 400031 storage/store_snapshot.go:775  [s1,r1/1:/M{in-ax}] sending preemptive snapshot d58c59b1 at applied index 21
I190402 05:44:02.925996 400031 storage/store_snapshot.go:818  [s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 60, log entries: 0, rate-limit: 8.0 MiB/sec, 0.01s
I190402 05:44:02.927084 401640 storage/replica_raftstorage.go:814  [s4,r1/?:{-}] applying preemptive snapshot at index 21 (id=d58c59b1, encoded size=8515, 1 rocksdb batches, 0 log entries)
I190402 05:44:02.927478 401640 storage/replica_raftstorage.go:820  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190402 05:44:02.928292 400031 storage/replica_command.go:796  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0]
I190402 05:44:02.930950 400031 storage/replica_raft.go:394  [s1,r1/1:/M{in-ax},txn=84926cba] proposing ADD_REPLICA((n4,s4):4): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3 (n4,s4):4] next=5
I190402 05:44:03.085903 400031 storage/replica_command.go:796  [s2,r1/2:/M{in-ax}] change replicas (REMOVE_REPLICA (n1,s1):1): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=0]
I190402 05:44:03.089400 400031 storage/replica_raft.go:394  [s2,r1/2:/M{in-ax},txn=9eb320ed] proposing REMOVE_REPLICA((n1,s1):1): updated=[(n4,s4):4 (n2,s2):2 (n3,s3):3] next=5
I190402 05:44:03.169579 402250 storage/store.go:2479  [replicaGC,s1,r1/1:/M{in-ax}] removing replica r1/1
I190402 05:44:03.179876 402250 storage/replica_destroy.go:150  [replicaGC,s1,r1/1:/M{in-ax}] removed 53 (49+4) keys in 10ms [clear=0ms commit=10ms]
I190402 05:44:04.354628 401110 gossip/gossip.go:1510  [n3] node has connected to cluster via gossip
I190402 05:44:04.968892 401304 gossip/gossip.go:1510  [n4] node has connected to cluster via gossip
I190402 05:44:05.032628 400154 gossip/gossip.go:1510  [n2] node has connected to cluster via gossip
I190402 05:44:18.189797 400114 storage/compactor/compactor.go:329  [s1,compactor] purging suggested compaction for range /Min - /Max that contains live data
I190402 05:44:48.096059 585749 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190402 05:44:48.096098 585749 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190402 05:44:48.096124 585666 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190402 05:44:48.096138 585666 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190402 05:44:48.096163 585747 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190402 05:44:48.096177 585747 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190402 05:44:48.096204 585748 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190402 05:44:48.096223 585748 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
W190402 05:44:48.097322 400154 gossip/gossip.go:1496  [n2] no incoming or outgoing connections
--- FAIL: TestReplicateRemovedNodeDisruptiveElection (45.58s)
    client_raft_test.go:3285: did not get expected ReplicaTooOldError error

```

SHA: https://github.com/cockroachdb/cockroach/commits/b1d5ae29d52698485b5398b57178c2ee02b3d49a

Parameters:
```
TAGS=
GOFLAGS=-parallel=4
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestReplicateRemovedNodeDisruptiveElection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1246248&tab=buildLog

```
I190417 12:57:11.529581 449346 storage/replica_raftstorage.go:820  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190417 12:57:11.530337 448223 storage/replica_command.go:792  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I190417 12:57:11.533572 448223 storage/replica_raft.go:396  [s1,r1/1:/M{in-ax},txn=ff141b20] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I190417 12:57:11.543637 448223 storage/store_snapshot.go:775  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 64a3cd99 at applied index 21
I190417 12:57:11.543893 448223 storage/store_snapshot.go:818  [s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 60, log entries: 0, rate-limit: 8.0 MiB/sec, 0.00s
I190417 12:57:11.544318 449641 storage/replica_raftstorage.go:814  [s4,r1/?:{-}] applying preemptive snapshot at index 21 (id=64a3cd99, encoded size=8515, 1 rocksdb batches, 0 log entries)
I190417 12:57:11.545096 449641 storage/replica_raftstorage.go:820  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=1ms]
I190417 12:57:11.545452 448223 storage/replica_command.go:792  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0]
I190417 12:57:11.547177 448223 storage/replica_raft.go:396  [s1,r1/1:/M{in-ax},txn=c386450b] proposing ADD_REPLICA((n4,s4):4): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3 (n4,s4):4] next=5
I190417 12:57:11.793472 448223 storage/replica_command.go:792  [s2,r1/2:/M{in-ax}] change replicas (REMOVE_REPLICA (n1,s1):1): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=0]
I190417 12:57:11.795420 448223 storage/replica_raft.go:396  [s2,r1/2:/M{in-ax},txn=112e2c14] proposing REMOVE_REPLICA((n1,s1):1): updated=[(n4,s4):4 (n2,s2):2 (n3,s3):3] next=5
I190417 12:57:11.799012 451748 storage/store.go:2520  [replicaGC,s1,r1/1:/M{in-ax}] removing replica r1/1
I190417 12:57:11.799416 451748 storage/replica_destroy.go:150  [replicaGC,s1,r1/1:/M{in-ax}] removed 53 (49+4) keys in 0ms [clear=0ms commit=0ms]
I190417 12:57:12.973670 448856 gossip/gossip.go:1510  [n3] node has connected to cluster via gossip
I190417 12:57:13.218299 448416 gossip/gossip.go:1510  [n2] node has connected to cluster via gossip
I190417 12:57:13.985919 449401 gossip/gossip.go:1510  [n4] node has connected to cluster via gossip
I190417 12:57:26.799737 448370 storage/compactor/compactor.go:329  [s1,compactor] purging suggested compaction for range /Min - /Max that contains live data
I190417 12:57:56.799157 813516 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190417 12:57:56.799160 813514 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190417 12:57:56.799198 813516 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_gc_batcher
I190417 12:57:56.799234 813517 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190417 12:57:56.799264 813517 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_gc_batcher
I190417 12:57:56.799240 813515 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
W190417 12:57:56.800128 449588 storage/raft_transport.go:583  while processing outgoing Raft queue to node 2: rpc error: code = Unavailable desc = transport is closing:
W190417 12:57:56.800163 451689 storage/raft_transport.go:583  while processing outgoing Raft queue to node 2: rpc error: code = Unavailable desc = transport is closing:
--- FAIL: TestReplicateRemovedNodeDisruptiveElection (45.42s)
    client_raft_test.go:3285: did not get expected ReplicaTooOldError error

```

SHA: https://github.com/cockroachdb/cockroach/commits/83de585d331b05a4aa02a65b353bed6bf829b696

Parameters:
```
TAGS=
GOFLAGS=-parallel=4
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestReplicateRemovedNodeDisruptiveElection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1248040&tab=buildLog

```
I190418 06:48:25.725823 339612 storage/replica_raftstorage.go:820  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190418 06:48:25.726839 338811 storage/replica_command.go:792  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I190418 06:48:25.732339 338811 storage/replica_raft.go:396  [s1,r1/1:/M{in-ax},txn=9a8b6afa] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I190418 06:48:25.734412 338811 storage/store_snapshot.go:775  [s1,r1/1:/M{in-ax}] sending preemptive snapshot ac497fa1 at applied index 21
I190418 06:48:25.734676 338811 storage/store_snapshot.go:818  [s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 60, log entries: 0, rate-limit: 8.0 MiB/sec, 0.00s
I190418 06:48:25.735654 339742 storage/replica_raftstorage.go:814  [s4,r1/?:{-}] applying preemptive snapshot at index 21 (id=ac497fa1, encoded size=8515, 1 rocksdb batches, 0 log entries)
I190418 06:48:25.736015 339742 storage/replica_raftstorage.go:820  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190418 06:48:25.736461 338811 storage/replica_command.go:792  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0]
I190418 06:48:25.739670 338811 storage/replica_raft.go:396  [s1,r1/1:/M{in-ax},txn=4156909f] proposing ADD_REPLICA((n4,s4):4): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3 (n4,s4):4] next=5
I190418 06:48:25.886606 338811 storage/replica_command.go:792  [s2,r1/2:/M{in-ax}] change replicas (REMOVE_REPLICA (n1,s1):1): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=0]
I190418 06:48:25.889099 338811 storage/replica_raft.go:396  [s2,r1/2:/M{in-ax},txn=d49b5daf] proposing REMOVE_REPLICA((n1,s1):1): updated=[(n4,s4):4 (n2,s2):2 (n3,s3):3] next=5
I190418 06:48:25.895283 340679 storage/store.go:2520  [replicaGC,s1,r1/1:/M{in-ax}] removing replica r1/1
I190418 06:48:25.896343 340679 storage/replica_destroy.go:150  [replicaGC,s1,r1/1:/M{in-ax}] removed 53 (49+4) keys in 1ms [clear=0ms commit=1ms]
I190418 06:48:27.220816 339258 gossip/gossip.go:1510  [n3] node has connected to cluster via gossip
I190418 06:48:27.766571 339602 gossip/gossip.go:1510  [n4] node has connected to cluster via gossip
I190418 06:48:27.800719 339079 gossip/gossip.go:1510  [n2] node has connected to cluster via gossip
I190418 06:48:40.897123 338910 storage/compactor/compactor.go:329  [s1,compactor] purging suggested compaction for range /Min - /Max that contains live data
I190418 06:49:10.893171 601821 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190418 06:49:10.893207 601820 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190418 06:49:10.893222 601819 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190418 06:49:10.893234 601820 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_gc_batcher
I190418 06:49:10.893213 601821 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_gc_batcher
I190418 06:49:10.893227 601818 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190418 06:49:10.893247 601819 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_gc_batcher
W190418 06:49:10.894053 339602 gossip/gossip.go:1496  [n4] no incoming or outgoing connections
--- FAIL: TestReplicateRemovedNodeDisruptiveElection (45.33s)
    client_raft_test.go:3285: did not get expected ReplicaTooOldError error

```

SHA: https://github.com/cockroachdb/cockroach/commits/826098540768fcb41e304fd114a981bb147561cf

Parameters:
```
TAGS=
GOFLAGS=-parallel=4
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestReplicateRemovedNodeDisruptiveElection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1264191&tab=buildLog

```
I190427 05:46:29.948251 458100 storage/replica_command.go:792  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I190427 05:46:29.950875 458100 storage/replica_raft.go:396  [s1,r1/1:/M{in-ax},txn=c3fe17c7] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I190427 05:46:29.952698 458100 storage/store_snapshot.go:775  [s1,r1/1:/M{in-ax}] sending preemptive snapshot f68827cc at applied index 21
I190427 05:46:29.952964 458100 storage/store_snapshot.go:818  [s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 60, log entries: 0, rate-limit: 8.0 MiB/sec, 0.00s
I190427 05:46:29.953425 459289 storage/replica_raftstorage.go:814  [s4,r1/?:{-}] applying preemptive snapshot at index 21 (id=f68827cc, encoded size=8661, 1 rocksdb batches, 0 log entries)
I190427 05:46:29.953695 459289 storage/replica_raftstorage.go:820  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190427 05:46:29.954210 458100 storage/replica_command.go:792  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0]
I190427 05:46:29.956362 458100 storage/replica_raft.go:396  [s1,r1/1:/M{in-ax},txn=34487b6b] proposing ADD_REPLICA((n4,s4):4): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3 (n4,s4):4] next=5
I190427 05:46:30.105965 458100 storage/replica_command.go:792  [s2,r1/2:/M{in-ax}] change replicas (REMOVE_REPLICA (n1,s1):1): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=0]
I190427 05:46:30.108691 458100 storage/replica_raft.go:396  [s2,r1/2:/M{in-ax},txn=f23742b2] proposing REMOVE_REPLICA((n1,s1):1): updated=[(n4,s4):4 (n2,s2):2 (n3,s3):3] next=5
I190427 05:46:30.111881 460399 storage/store.go:2520  [replicaGC,s1,r1/1:/M{in-ax}] removing replica r1/1
I190427 05:46:30.112634 460399 storage/replica_destroy.go:150  [replicaGC,s1,r1/1:/M{in-ax}] removed 53 (49+4) keys in 0ms [clear=0ms commit=0ms]
I190427 05:46:31.771975 458206 gossip/gossip.go:1510  [n2] node has connected to cluster via gossip
I190427 05:46:31.784670 458648 gossip/gossip.go:1510  [n4] node has connected to cluster via gossip
I190427 05:46:32.076163 458333 gossip/gossip.go:1510  [n3] node has connected to cluster via gossip
I190427 05:46:45.113031 458106 storage/compactor/compactor.go:329  [s1,compactor] purging suggested compaction for range /Min - /Max that contains live data
I190427 05:47:15.111761 922072 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190427 05:47:15.111792 922072 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190427 05:47:15.111817 922069 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190427 05:47:15.111833 922069 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190427 05:47:15.111855 922070 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190427 05:47:15.111868 922070 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190427 05:47:15.111885 922071 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190427 05:47:15.111900 922071 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
W190427 05:47:15.113159 458206 gossip/gossip.go:1496  [n2] no incoming or outgoing connections
W190427 05:47:15.113289 458333 gossip/gossip.go:1496  [n3] no incoming or outgoing connections
--- FAIL: TestReplicateRemovedNodeDisruptiveElection (45.26s)
    client_raft_test.go:3469: did not get expected ReplicaTooOldError error

```

SHA: https://github.com/cockroachdb/cockroach/commits/bb50fd396f6ce79258b744e1f8efa2d1bc9dfbd2

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestReplicateRemovedNodeDisruptiveElection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1273055&tab=buildLog

```
I190503 06:16:55.006532 755218 storage/replica_raftstorage.go:820  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190503 06:16:55.007319 753956 storage/replica_command.go:797  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I190503 06:16:55.013582 753956 storage/replica_raft.go:396  [s1,r1/1:/M{in-ax},txn=a66308df] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I190503 06:16:55.018476 753956 storage/store_snapshot.go:780  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 04484add at applied index 21
I190503 06:16:55.018719 753956 storage/store_snapshot.go:823  [s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 60, log entries: 0, rate-limit: 8.0 MiB/sec, 0.00s
I190503 06:16:55.019049 755335 storage/replica_raftstorage.go:814  [s4,r1/?:{-}] applying preemptive snapshot at index 21 (id=04484add, encoded size=8699, 1 rocksdb batches, 0 log entries)
I190503 06:16:55.019351 755335 storage/replica_raftstorage.go:820  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190503 06:16:55.021274 753956 storage/replica_command.go:797  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0]
I190503 06:16:55.025028 753956 storage/replica_raft.go:396  [s1,r1/1:/M{in-ax},txn=72e95cc3] proposing ADD_REPLICA((n4,s4):4): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3 (n4,s4):4] next=5
I190503 06:16:55.101931 753956 storage/replica_command.go:797  [s2,r1/2:/M{in-ax}] change replicas (REMOVE_REPLICA (n1,s1):1): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=0]
I190503 06:16:55.104788 753956 storage/replica_raft.go:396  [s2,r1/2:/M{in-ax},txn=50f1e5ed] proposing REMOVE_REPLICA((n1,s1):1): updated=[(n4,s4):4 (n2,s2):2 (n3,s3):3] next=5
I190503 06:16:55.115046 755805 storage/store.go:2507  [replicaGC,s1,r1/1:/M{in-ax}] removing replica r1/1
I190503 06:16:55.115442 755805 storage/replica_destroy.go:150  [replicaGC,s1,r1/1:/M{in-ax}] removed 53 (49+4) keys in 0ms [clear=0ms commit=0ms]
I190503 06:16:56.593599 754143 gossip/gossip.go:1510  [n2] node has connected to cluster via gossip
I190503 06:16:57.235674 754416 gossip/gossip.go:1510  [n3] node has connected to cluster via gossip
I190503 06:16:57.395581 754689 gossip/gossip.go:1510  [n4] node has connected to cluster via gossip
I190503 06:17:10.115761 754119 storage/compactor/compactor.go:329  [s1,compactor] purging suggested compaction for range /Min - /Max that contains live data
I190503 06:17:40.115346 997876 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190503 06:17:40.115375 997876 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190503 06:17:40.115397 997825 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190503 06:17:40.115411 997825 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190503 06:17:40.115431 997826 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190503 06:17:40.115445 997826 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190503 06:17:40.115465 997875 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190503 06:17:40.115477 997875 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
W190503 06:17:40.116837 754416 gossip/gossip.go:1496  [n3] no incoming or outgoing connections
--- FAIL: TestReplicateRemovedNodeDisruptiveElection (45.28s)
    client_raft_test.go:3469: did not get expected ReplicaTooOldError error

```

SHA: https://github.com/cockroachdb/cockroach/commits/72d089cfc1a0b67c299439d328587345f1125b74

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestReplicateRemovedNodeDisruptiveElection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1277914&tab=buildLog

```
I190507 07:14:48.011888 532420 storage/replica_raft.go:396  [s1,r1/1:/M{in-ax},txn=ad9f5f7a] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I190507 07:14:48.013681 532420 storage/store_snapshot.go:780  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 29ba3ced at applied index 21
I190507 07:14:48.013904 532420 storage/store_snapshot.go:823  [s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 60, log entries: 0, rate-limit: 8.0 MiB/sec, 0.00s
I190507 07:14:48.014326 533549 storage/replica_raftstorage.go:814  [s4,r1/?:{-}] applying preemptive snapshot at index 21 (id=29ba3ced, encoded size=8699, 1 rocksdb batches, 0 log entries)
I190507 07:14:48.014701 533549 storage/replica_raftstorage.go:820  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190507 07:14:48.015492 532420 storage/replica_command.go:797  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0]
I190507 07:14:48.021840 532420 storage/replica_raft.go:396  [s1,r1/1:/M{in-ax},txn=9a725e74] proposing ADD_REPLICA((n4,s4):4): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3 (n4,s4):4] next=5
I190507 07:14:48.183777 532420 storage/replica_command.go:797  [s2,r1/2:/M{in-ax}] change replicas (REMOVE_REPLICA (n1,s1):1): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=0]
I190507 07:14:48.187553 532420 storage/replica_raft.go:396  [s2,r1/2:/M{in-ax},txn=89b999bb] proposing REMOVE_REPLICA((n1,s1):1): updated=[(n4,s4):4 (n2,s2):2 (n3,s3):3] next=5
I190507 07:14:48.200186 534102 storage/store.go:2507  [replicaGC,s1,r1/1:/M{in-ax}] removing replica r1/1
I190507 07:14:48.200853 534102 storage/replica_destroy.go:150  [replicaGC,s1,r1/1:/M{in-ax}] removed 53 (49+4) keys in 0ms [clear=0ms commit=0ms]
I190507 07:14:49.323462 532350 gossip/gossip.go:1510  [n1] node has connected to cluster via gossip
I190507 07:14:49.505548 532524 gossip/gossip.go:1510  [n2] node has connected to cluster via gossip
I190507 07:14:49.871725 532984 gossip/gossip.go:1510  [n3] node has connected to cluster via gossip
I190507 07:14:49.973434 533150 gossip/gossip.go:1510  [n4] node has connected to cluster via gossip
I190507 07:15:03.201557 532503 storage/compactor/compactor.go:329  [s1,compactor] purging suggested compaction for range /Min - /Max that contains live data
I190507 07:15:33.192367 726208 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190507 07:15:33.192409 726208 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190507 07:15:33.192468 726207 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190507 07:15:33.192498 726207 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190507 07:15:33.192553 726205 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190507 07:15:33.192573 726205 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190507 07:15:33.192593 726206 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190507 07:15:33.192611 726206 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
W190507 07:15:33.193612 533538 storage/raft_transport.go:583  while processing outgoing Raft queue to node 3: rpc error: code = Unavailable desc = transport is closing:
W190507 07:15:33.193882 533150 gossip/gossip.go:1496  [n4] no incoming or outgoing connections
--- FAIL: TestReplicateRemovedNodeDisruptiveElection (45.46s)
    client_raft_test.go:3469: did not get expected ReplicaTooOldError error

```

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -3648,6 +3648,7 @@ func (s *Store) processRequestQueue(ctx context.Context, rangeID roachpb.RangeID
 	q.Unlock()
 
 	var lastRepl *Replica
+	var hadError bool
 	for i, info := range infos {
 		last := i == len(infos)-1
 		pErr := s.withReplicaForRequest(info.respStream.Context(), info.req,
@@ -3669,21 +3670,32 @@ func (s *Store) processRequestQueue(ctx context.Context, rangeID roachpb.RangeID
 				return pErr
 			})
 		if pErr != nil {
-			// If we're unable to process the request, clear the request queue. This
-			// only happens if we couldn't create the replica because the request was
-			// targeted to a removed range. This is also racy and could cause us to
-			// drop messages to the deleted range occasionally (#18355), but raft
-			// will just retry.
-			q.Lock()
-			if len(q.infos) == 0 {
-				s.replicaQueues.Delete(int64(rangeID))
-			}
-			q.Unlock()
+			hadError = true
 			if err := info.respStream.Send(newRaftMessageResponse(info.req, pErr)); err != nil {
 				// Seems excessive to log this on every occurrence as the other side
 				// might have closed.
 				log.VEventf(ctx, 1, "error sending error: %s", err)
 			}
+		}
+	}
+
+	if hadError {
+		// If we're unable to process a request, consider dropping the request queue
+		// to free up space in the map.
+		// This is relevant if requests failed because the target replica could not
+		// be created (for example due to the Raft tombstone). The particular code
+		// here takes into account that we don't want to drop the queue if there
+		// are other messages waiting on it, or if the target replica exists. Raft
+		// tolerates the occasional dropped message, but our unit tests are less
+		// forgiving.
+		//
+		// See https://github.com/cockroachdb/cockroach/issues/30951#issuecomment-428010411.
+		if _, exists := s.mu.replicas.Load(int64(rangeID)); !exists {
+			q.Lock()
+			if len(q.infos) == 0 {
+				s.replicaQueues.Delete(int64(rangeID))
+			}
+			q.Unlock()
 		}
 	}
 

